### PR TITLE
Clone crates.io-index instead of querying it through GitHub's API

### LINKED ISSUE #66: Local registry index cache
For now, we get latest version of crate via HTTP call, each crate cost a HTTP call, which might be expensive.

If we keep a full index local cache, and timing sync it, that might be cheaper and faster.

Here are my ideas:

1. Download github archive file every 5 mins and load everything in memory, for now, crates.io-index archive file is 35MB.
2. Clone and read via `git2` crate, incremental synchronization every 1 min.
3. Download archive file when first launch, incremental synchronization via [github api](https://docs.github.com/en/free-pro-team@latest/rest/reference/git#get-a-tree)

### PR DESCRIPTION
**Pull request by paolobarbolini**
Closes #66

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -70,7 +71,13 @@ async fn main() {
 
     let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), port);
 
-    let mut engine = Engine::new(client.clone(), logger.new(o!()));
+    let mut managed_index = ManagedIndex::new(Duration::from_secs(20));
+    let index = managed_index.index();
+    tokio::spawn(async move {
+        managed_index.refresh_at_interval().await;
+    });
```
Comment (robjtede) [inline src/main.rs:78]: pretty sure thats redundant

```suggestion
    tokio::spawn(managed_index.refresh_at_interval());
```

**Comment by robjtede** [comment]
Haven't run the code yet but we should be able to understand the behavior in these situations:
1. index not yet fetched (empty disk)
1. index loaded (easy case)
1. index refreshing/updating
1. GitHub down or other outbound network error when there is no cache
1. GitHub down or other outbound network error when the cache is loaded

I think in the case 1 it's perfectly fine to block server starting even until the initial index load is done. It's pretty easy using Kubernetes to only start serving requests after that using a readiness condition. Other cases should be handled with more consideration to having some reasonable, defined response that doesn't stop requests being served.

File: src/main.rs (hunk 1 above)
Comment (paolobarbolini) [reply src/main.rs:78]: It doesn't work because it's not taking ownership of `managed_index`.

**Comment by paolobarbolini** [comment]
> I think in the case 1 it's perfectly fine to block server starting even until the initial index load is done

:+1: That would work if we end-up staying with Kubernetes

**Comment by paolobarbolini** [comment]
I've merged this branch locally a few weeks ago when I started hosting deps.rs on my VPS, and it's been working well.

Is there anything that should be changed? Can we get an approval to merge?

**Review by DCjanus — APPROVED** [review]
LGTM

File: src/main.rs [hunk 2]
```diff
@@ -70,7 +71,15 @@ async fn main() {
 
     let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), port);
 
-    let mut engine = Engine::new(client.clone(), logger.new(o!()));
+    let mut managed_index = ManagedIndex::new(Duration::from_secs(20), logger.clone());
+    let _ = managed_index.initial_clone().await;
```
[the remark this answers is not in the document]
Comment (paolobarbolini) [reply src/main.rs]: At the moment it's just prints the error. The clone operation will be retried by `refresh_at_interval` so it shouldn't be a problem unless something had been configured very wrongly and the operation can never succeed.

**Pull request merged**